### PR TITLE
Upgrade Examples and tests to use Helm 3

### LINKED ISSUE #82: Use Helm 3/remove Tiller?
Thanks so much for providing [this article](https://cloud.google.com/blog/products/devops-sre/deploying-a-production-grade-helm-release-on-gke-with-terraform) and the accompanying repo! Our team needs to deploy a single Helm chart in GKE with as few moving parts as possible to help us with federal compliance, and this is very instructive.

In the time since the article was written, [Helm 3 dropped the use of Tiller](https://helm.sh/docs/faq/#removal-of-tiller), further simplifying and integrating with k8s. Do you have any plans to update your repo to make use of Helm 3 and drop the Tiller component?

### PR DESCRIPTION
**Pull request by autero1**
This PR updates the tests and examples to use Helm 3 instead of Helm 2 with Tiller. As a result, one of the examples was removed as redundant and the new `gke-basic-helm` was moved to the root of the repo.

The `TestGKEBasicHelm` test case has some redundancy, as both the Terraform example and the test deploy a chart, but I decided to leave it that way - primarily because I felt it made the example itself much better, because the user actually gets to interact with the cluster.

Tests pass locally, the only question mark for me is how they behave in CircleCI. 

Fixes #82

### CODE REVIEW

**Comment by autero1** [comment]
Looks like the tests failed. I'll try and see tomorrow what the problem was.

File: .circleci/config.yml [hunk 1]
```diff
@@ -3,13 +3,12 @@ defaults: &defaults
   environment:
     GRUNTWORK_INSTALLER_VERSION: v0.0.21
     TERRATEST_LOG_PARSER_VERSION: v0.13.13
-    KUBERGRUNT_VERSION: v0.3.9
-    HELM_VERSION: v2.11.0
+    HELM_VERSION: v3.1.0
```
Comment (robmorgan) [inline .circleci/config.yml:6]: I might update this to `v3.1.2` to match `terraform-aws-eks`.

**Comment by autero1** [comment]
> My only question is should the default example in the root of the repo be a private GKE cluster? I believe that's what we had originally?

The original root example was the private Tiller example. That's why I opted for the Helm example as the new root.

**Comment by robmorgan** [comment]
@autero1 sure but is it possible to do a Helm example in the new root with a private cluster?

**Comment by autero1** [comment]
Ah, you're right... 😄 It was indeed a private cluster. I'll change that and fix the tests.

File: examples/gke-basic-helm/README.md [hunk 2]
```diff
@@ -0,0 +1,88 @@
+# GKE Basic Helm Example
+
+The root folder contains an example of how to deploy a GKE Public Cluster with and an example chart
```
Comment (zackproser) [inline examples/gke-basic-helm/README.md:3]:
```suggestion
The root folder contains an example of how to deploy a GKE Public Cluster with an example chart 
```

File: examples/gke-basic-helm/README.md [hunk 3]
```diff
@@ -0,0 +1,88 @@
...
+
+If you wish to access the deployed service, you use the `kubectl port-forward` command to forward a local port to the deployed service:
+
+```
+❯ kubectl port-forward deployment/nginx 8080:8080 -n default
+Forwarding from 127.0.0.1:8080 -> 8080
+Forwarding from [::1]:8080 -> 8080
+```
+
+You can now access the deployed service by opening your web browser to `http://localhost:8088`.
```
Comment (zackproser) [inline examples/gke-basic-helm/README.md:84]: Did you mean for this to be `http:localhost:8080` ?

File: examples/gke-basic-helm/README.md [hunk 4]
```diff
@@ -0,0 +1,88 @@
+# GKE Basic Helm Example
```
Comment (zackproser) [inline examples/gke-basic-helm/README.md:1]: Awesome docs :+1:

File: main.tf [hunk 5]
```diff
@@ -262,136 +231,27 @@ resource "kubernetes_cluster_role_binding" "user" {
...
-  tiller_tls_secret_name                   = local.tls_secret_name
-  namespace                                = local.tiller_namespace
-  tiller_image_version                     = local.tiller_version
-
-  # Kubergrunt will store the private key under the key "tls.pem" in the corresponding Secret resource, which will be
-  # accessed as a file when mounted into the container.
-  tiller_tls_key_file_name = "tls.pem"
-
-  dependencies = [null_resource.tiller_tls_certs.id, kubernetes_cluster_role_binding.user.id]
+data "helm_repository" "bitnami" {
```
Comment (zackproser) [inline main.tf:245]: NIT: maybe worth a comment about what is happening here - would probably be helpful for folks new to Helm concepts.

Comment (zackproser) [reply main.tf:245]: Specifically that its going to look up a chart from a chart repository.

File: examples/gke-basic-helm/README.md (hunk 3 above)
Comment (autero1) [reply examples/gke-basic-helm/README.md:84]: Good catch!

File: main.tf (hunk 5 above)
Comment (yorinasub17) [reply main.tf:269]: NIT: There is actually a critical issue with this data source (https://github.com/terraform-providers/terraform-provider-helm/issues/438 and https://github.com/terraform-providers/terraform-provider-helm/issues/416), and they are actually considering removing it entirely. I've found that it is much more robust and stable to directly use the repo URL in the `repository` attribute of the `helm_release` resource instead.

Comment (autero1) [reply main.tf:269]: Thanks @yorinasub17 for pointing this out. I'll fix this!

**Comment by autero1** [comment]
I bumped a lot of tool versions, but don't know if it fixed the test issue.

A more confusing thing was an issue with the network module. Since v0.3.0 the `terraform-google-network` changed NAT routing from public to private. I used v0.4.0 in the example (other examples untouched) and I initially failed to install any charts, because the nodes did not have internet connectivity. 

After some experimentation, the example is successfully deployed to the `private` subnet, with `private` network tag, so the cluster is truly private. 

Decided to not change the network module version for the other example to not bloat the PR even further.

File: main.tf [hunk 6]
```diff
@@ -1,6 +1,7 @@
 # ---------------------------------------------------------------------------------------------------------------------
-# DEPLOY A GKE PRIVATE CLUSTER W/ TILLER IN GOOGLE CLOUD PLATFORM
-# This is an example of how to use the gke-cluster module to deploy a private Kubernetes cluster in GCP
+# DEPLOY A GKE PUBLIC CLUSTER IN GOOGLE CLOUD PLATFORM
```
Comment (robmorgan) [inline main.tf:2]:
```suggestion
# DEPLOY A GKE PRIVATE CLUSTER IN GOOGLE CLOUD PLATFORM
```

File: main.tf [hunk 7]
```diff
@@ -1,6 +1,7 @@
...
+# This is an example of how to use the gke-cluster module to deploy a public Kubernetes cluster in GCP with a
```
Comment (robmorgan) [inline main.tf:3]:
```suggestion
# This is an example of how to use the gke-cluster module to deploy a private Kubernetes cluster in GCP.
```

File: main.tf [hunk 8]
```diff
@@ -1,6 +1,7 @@
...
+# Load Balancer in front of it.
```
Comment (robmorgan) [inline main.tf:4]:
```suggestion
```

File: main.tf [hunk 9]
```diff
@@ -92,10 +91,10 @@ module "gke_cluster" {
   location = var.location
   network  = module.vpc_network.network
 
-  # We're deploying the cluster in the 'public' subnetwork to allow outbound internet access
+  # Deploy the cluster in the 'private' subnetwork, outbound internet access will be provided by NAT
   # See the network access tier table for full details:
   # https://github.com/gruntwork-io/terraform-google-network/tree/master/modules/vpc-network#access-tier
-  subnetwork = module.vpc_network.public_subnetwork
+  subnetwork = module.vpc_network.private_subnetwork
```
Comment (robmorgan) [inline main.tf:97]: nice 👍

File: test/go.mod [hunk 10]
```diff
@@ -0,0 +1,8 @@
+module github.com/gruntwork-io/terraform-google-gke/test
+
+go 1.14
```
Comment (robmorgan) [inline test/go.mod:3]:
```suggestion
go 1.13
```
Use CircleCI infrastructure.

**Comment by autero1** [comment]
> Private Subnetworks with NAT routing enabled so the Nodes have internet connectivity.
> A Private GKE Cluster with a Helm chart deployed.

That is correct. And I agree that with the fix in the network module, we now have correct behaviour.

I applied the fixes you suggested.

**Comment by robmorgan** [comment]
@autero1 cool man, I'll kick off the tests again

**Comment by robmorgan** [comment]
@autero1 It looks like they passed in https://circleci.com/gh/gruntwork-io/terraform-google-gke/457. Ok to merge?

**Comment by autero1** [comment]
Ok to merge! Though the tests seemed to fail until you made final tweaks to the PR branch. Will you merge those after you merge this PR?

**Comment by robmorgan** [comment]
@autero1 yep, I'll merge my branch. It will take all of your commits, apply mine on top and close this PR!

**Pull request merged**